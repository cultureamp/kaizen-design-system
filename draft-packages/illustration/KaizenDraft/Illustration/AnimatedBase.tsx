import React, { useEffect, useRef, useState } from "react"
import { assetUrl } from "@kaizen/hosted-assets"
import lottie from "lottie-web"
import { BaseProps } from "./Base"
import { fetchPath } from "./utils"
import styles from "./style.module.scss"

export type AnimatedBaseProps = {
  autoplay?: boolean
  loop?: boolean
  isAnimated?: boolean
}

// @TODO
// - accessibility (maybe using figure since I can't add metadata to the lottie animation?)
// - fallback to static image
// - error handling when asset loading fails
// - allow delay on animation
export const AnimatedBase = ({
  name,
  autoplay = true,
  loop = false,
  alt,
}: AnimatedBaseProps & BaseProps) => {
  const lottiePlayer = useRef<HTMLDivElement>(null)
  const [playerLoaded, setPlayerLoaded] = useState<boolean>(false)
  const [asset, setAsset] = React.useState<null | any>(null)

  React.useEffect(() => {
    let didCancel = false;
    const fetchData = async () => {
      const path = assetUrl(name)
      const srcParsed = await fetchPath(path)
      if (!didCancel) {
        setAsset(srcParsed)
      }
    }
    fetchData()
    return () => {
      didCancel = true;
    };
  }, [])

  useEffect(() => {
    const initialiseLottiePlayer = () => {
      if (asset && lottiePlayer.current !== null && !playerLoaded) {
        setPlayerLoaded(true)
        lottie.loadAnimation({
          container: lottiePlayer.current,
          renderer: "svg",
          loop,
          autoplay,
          animationData: asset
        })
      }
    }
    initialiseLottiePlayer()
  }, [asset])

  return (
    <div className={styles.wrapper}>
      {/* Avoid jump when asset loads */}
      {!playerLoaded && (
        <svg data-testid="loading" viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg">
          <circle fill="#f9f9f9" cx="48" cy="48" r="48" />
        </svg>
      )}
      <div data-testid="lottie-player" ref={lottiePlayer} />
    </div>
  )
}
