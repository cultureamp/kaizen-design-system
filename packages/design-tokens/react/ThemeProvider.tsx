import React from "react"
import { ThemeManager, defaultTheme, Theme } from "../"

export const ThemeContext = React.createContext<Theme>(defaultTheme)

/**
 * Wrap your application in this provider using a ThemeManager, to synchronise it with a react context.
 * This allows child react elements to more easily use theme variables, using the {@link useTheme} hook.
 */
export const ThemeProvider = ({
  themeManager,
  ...props
}: {
  themeManager: ThemeManager
  children: React.ReactNode
}) => {
  const [theme, setTheme] = React.useState<Theme>(
    themeManager.getCurrentTheme()
  )
  React.useEffect(() => {
    let cancelled = false
    const listener = (newTheme: Theme) => {
      if (!cancelled) setTheme(newTheme)
    }
    themeManager.addThemeChangeListener(listener)
    return () => {
      cancelled = true
      themeManager.removeThemeChangeListener(listener)
    }
  }, [])

  return (
    <ThemeContext.Provider value={theme}>
      {props.children}
    </ThemeContext.Provider>
  )
}

export const useTheme = () => React.useContext(ThemeContext)
