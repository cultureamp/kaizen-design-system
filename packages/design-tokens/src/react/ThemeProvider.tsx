import React, { useState } from "react"
import { ThemeManager, defaultTheme, Theme } from ".."

export const ThemeContext = React.createContext<Theme>(defaultTheme)

/**
 * Wrap your application in this provider using a ThemeManager, to synchronise it with a react context.
 * This allows child react elements to more easily use theme variables, using the {@link useTheme} hook.
 */
/**
 * @deprecated ThemeProvider is deprecated. Please use KaizenProvider from "@kaizen/components" instead.
 */
export const ThemeProvider = ({
  themeManager,
  ...props
}: {
  themeManager?: ThemeManager
  children: React.ReactNode
}): JSX.Element => {
  const [themeManagerInstance] = useState(
    () => themeManager || new ThemeManager(defaultTheme)
  )

  const [theme, setTheme] = React.useState<Theme>(
    themeManagerInstance.getCurrentTheme()
  )
  React.useEffect(() => {
    let cancelled = false
    const listener = (newTheme: Theme): void => {
      if (!cancelled) setTheme(newTheme)
    }
    themeManagerInstance.addThemeChangeListener(listener)
    return () => {
      cancelled = true
      themeManagerInstance.removeThemeChangeListener(listener)
    }
  }, [])

  return (
    <>
      <ThemeContext.Provider value={theme}>
        {props.children}
      </ThemeContext.Provider>
      <link
        rel="preload"
        href="https://d1e7r7b0lb8p4d.cloudfront.net/fonts/inter/inter-bold.woff2"
        as="font"
        type="font/woff2"
        crossOrigin="anonymous"
      />
      <link
        rel="preload"
        href="https://d1e7r7b0lb8p4d.cloudfront.net/fonts/inter/inter-demi-bold.woff2"
        as="font"
        type="font/woff2"
        crossOrigin="anonymous"
      />
      <link
        rel="preload"
        href="https://d1e7r7b0lb8p4d.cloudfront.net/fonts/inter/inter-medium.woff2"
        as="font"
        type="font/woff2"
        crossOrigin="anonymous"
      />
      <link
        rel="preload"
        href="https://d1e7r7b0lb8p4d.cloudfront.net/fonts/inter/inter-regular.woff2"
        as="font"
        type="font/woff2"
        crossOrigin="anonymous"
      />
      <link
        rel="preload"
        href="https://d1e7r7b0lb8p4d.cloudfront.net/fonts/tiempos/tiempos-headline-bold.woff2"
        as="font"
        type="font/woff2"
        crossOrigin="anonymous"
      />
    </>
  )
}

export const useTheme = (): Theme => React.useContext(ThemeContext)
