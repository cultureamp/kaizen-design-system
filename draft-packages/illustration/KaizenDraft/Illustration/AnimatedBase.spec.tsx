import { cleanup, render, screen, waitFor } from "@testing-library/react"
import * as React from "react"
import { AnimatedBase } from "./AnimatedBase"

afterEach(cleanup)

describe("<AnimatedBase />", () => {
  it("should render a fallback loading indicator", () => {
    render(<AnimatedBase name="" alt="" isAnimated/>)
    const loadingSvg = screen.getByTestId("loading")

    expect(loadingSvg).toMatchSnapshot()
    expect(loadingSvg).toBeTruthy()
  })

  // it("loads animation data into external lottie-plater", async () => {
  //   render(<AnimatedBase name="" alt="" isAnimated/>)
  //   await waitFor(() => screen.getByTestId("lottie-player"))
  // })
})
