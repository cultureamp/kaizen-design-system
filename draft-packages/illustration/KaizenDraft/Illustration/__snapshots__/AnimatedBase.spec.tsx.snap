// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnimatedBase /> should render a fallback indicator 1`] = `
<svg
  data-testid="loading"
  viewBox="0 0 100 100"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    cx="48"
    cy="48"
    fill="#f9f9f9"
    r="48"
  />
</svg>
`;

exports[`<AnimatedBase /> should render a fallback loading indicator 1`] = `
<svg
  data-testid="loading"
  viewBox="0 0 100 100"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    cx="48"
    cy="48"
    fill="#f9f9f9"
    r="48"
  />
</svg>
`;
